File: python/hot100/88.乘积最大子数组.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        maxp, minp, result = nums[0], nums[0], nums[0]
        n = len(nums)
        for i in range(1, n):
            num = nums[i]
            temp = maxp
            maxp = max(num, num * temp, num * minp)
            minp = min(num, num * temp, num * minp)
            result = max(maxp, result)
        return result

File: cpp/hot100/88.乘积最大子数组.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.empty()) return 0;
        long long maxp = nums[0], minp = nums[0], result = nums[0];
        int n = nums.size();
        for(int i = 1; i < n; ++i){
            long long num = nums[i];
            auto temp = maxp;
            maxp = max(num, max(num * temp, num * minp));
            minp = min(num, min(num * temp, num * minp));
            result = max(maxp, result);
        }
        return result;
    }
};